src/result_processor.c

```diff
@@ -845,3 +846,317 @@ ResultProcessor *RPCounter_New() {
+
+  // Free buffer memory blocks
+  array_foreach(bufferAndLocker->BufferBlocks, SearchResultsBlock, array_free(SearchResultsBlock));
+  array_free(bufferAndLocker->BufferBlocks);
+
+  rm_free(bufferAndLocker);
+}
+
+int rpbufferNext_bufferDocs(ResultProcessor *rp, SearchResult *res) {
+  RPBufferAndLocker *rpPufferAndLocker = (RPBufferAndLocker *)rp;
+
+  // Get the current index version so we can check later if any updates accured
+  // after we finish to buffer the results
+  RedisSearchCtx *RSctx = rpPufferAndLocker->base.parent->sctx;
+  double currentIndexVersion = IndexSpec_GetVersion(RSctx->spec);
+
+  // Keep fetching results from the upstream result processor until EOF is reached
+  int result_status;
+  SearchResult resToBuffer = {0};
+  SearchResult *CurrBlock = NULL;
+  // Get the next result and save it in the buffer
+  while((result_status = rp->upstream->Next(rp->upstream, &resToBuffer)) == RS_RESULT_OK) {
+
+    // Buffer the result.
+    CurrBlock = InsertResult(rpPufferAndLocker, &resToBuffer, CurrBlock);
+    
+    memset(&resToBuffer, 0, sizeof(SearchResult));
+    
+  }
+
+  // If we exit the loop because we got an error, or we have zero result, return without locking Redis.
+  if ((result_status != RS_RESULT_EOF &&
+      !(result_status == RS_RESULT_TIMEDOUT && RSGlobalConfig.timeoutPolicy == TimeoutPolicy_Return)) ||
+      IsBufferEmpty(rpPufferAndLocker)) {
+    return result_status;
+  }
+
+  // Now we have the data of all documents that pass the query filters,
+  // let's lock Redis to provide safe access to Redis keyspace 
+  
+  // unlockspec to avoid deadlocks
+  RedisSearchCtx_UnlockSpec(RSctx);
+
+  // Lock Redis to gurentee safe access to Redis keyspace
+  LockRedis(rpPufferAndLocker, RSctx->redisCtx);
+
+  // If the spec has been changed since we released the spec lock,
+  // we need to validate every buffered result
+  if (currentIndexVersion != IndexSpec_GetVersion(RSctx->spec)) {
+    rp->Next = rpbufferNext_ValidateAndYield;
+  } else { // Else we just return the results one by one
+    rp->Next = rpbufferNext_Yield;
+  }
+
+  // We don't lock the index spec because we assume that there 
+  // are no more access to the index down the pipeline and the data 
+  // we buffered remains valid.
+  return rp->Next(rp, res);
+}
+
+bool isResultValid(const SearchResult *res) {
+  // check if the doc is not marked deleted in the spec
+  return !(res->dmd->flags & Document_Deleted);
+}
+static void InvalidateBufferedResult(SearchResult *buffered_result) {
+  buffered_result->indexResult = NULL;
+  buffered_result->dmd = NULL;
+  buffered_result->scoreExplain = NULL;
+  memset(&buffered_result->rowdata, 0, sizeof(RLookupRow));
+}
+static void SetResult(SearchResult *buffered_result,  SearchResult *result_output) {
+  // Free the RLookup row before overriding it.
+  RLookupRow_Cleanup(&result_output->rowdata);
+  *result_output = *buffered_result;
+
+  InvalidateBufferedResult(buffered_result);
```

Comment:
is this phase relevant? the search result will no longer be in use so we can avoid cleaning it up with "nullify" everything